File: backend/src/main/java/com/backend/ord/exceptions/handlers/RESTExceptionHandler.java
package com.backend.ord.exceptions.handlers;

import com.backend.ord.exceptions.REST.*;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.ControllerAdvice;
import org.springframework.web.bind.annotation.ExceptionHandler;

class HTTPErrorResponse {
    private String message;
    private int status;

    HTTPErrorResponse(String message, int status) {
        this.message = message;
        this.status = status;
    }

    public static HTTPErrorResponseBuilder builder() {
        return new HTTPErrorResponseBuilder();
    }

    public String getMessage() {
        return this.message;
    }

    public int getStatus() {
        return this.status;
    }

    public void setMessage(String message) {
        this.message = message;
    }

    public void setStatus(int status) {
        this.status = status;
    }

    public boolean equals(final Object o) {
        if (o == this) return true;
        if (!(o instanceof HTTPErrorResponse)) return false;
        final HTTPErrorResponse other = (HTTPErrorResponse) o;
        if (!other.canEqual((Object) this)) return false;
        final Object this$message = this.getMessage();
        final Object other$message = other.getMessage();
        if (this$message == null ? other$message != null : !this$message.equals(other$message)) return false;
        if (this.getStatus() != other.getStatus()) return false;
        return true;
    }

    protected boolean canEqual(final Object other) {
        return other instanceof HTTPErrorResponse;
    }

    public int hashCode() {
        final int PRIME = 59;
        int result = 1;
        final Object $message = this.getMessage();
        result = result * PRIME + ($message == null ? 43 : $message.hashCode());
        result = result * PRIME + this.getStatus();
        return result;
    }

    public String toString() {
        return "HTTPErrorResponse(message=" + this.getMessage() + ", status=" + this.getStatus() + ")";
    }

    public static class HTTPErrorResponseBuilder {
        private String message;
        private int status;

        HTTPErrorResponseBuilder() {
        }

        public HTTPErrorResponseBuilder message(String message) {
            this.message = message;
            return this;
        }

        public HTTPErrorResponseBuilder status(int status) {
            this.status = status;
            return this;
        }

        public HTTPErrorResponse build() {
            return new HTTPErrorResponse(this.message, this.status);
        }

        public String toString() {
            return "HTTPErrorResponse.HTTPErrorResponseBuilder(message=" + this.message + ", status=" + this.status + ")";
        }
    }
}

@ControllerAdvice
public class RESTExceptionHandler {
    @ExceptionHandler(BadRequestException.class)
    public ResponseEntity<?> handleBadRequestException(BadRequestException e) {
        return handleException(e, 400);
    }

    @ExceptionHandler(UnauthorizedException.class)
    public ResponseEntity<?> handleUnauthorizedException(UnauthorizedException e) {
        return handleException(e, 401);
    }

    @ExceptionHandler(ForbiddenException.class)
    public ResponseEntity<?> handleForbiddenException(ForbiddenException e) {
        return handleException(e, 403);
    }

    @ExceptionHandler(NotFoundException.class)
    public ResponseEntity<?> handleNotFoundException(NotFoundException e) {
        return handleException(e, 404);
    }

    @ExceptionHandler(PayloadTooLargeException.class)
    public ResponseEntity<?> handlePayloadTooLargeException(PayloadTooLargeException e) {
        return handleException(e, 413);
    }

    @ExceptionHandler(InternalServerError.class)
    public ResponseEntity<?> handleInternalServerError(InternalServerError e) {
        return handleException(e, 500);
    }

    private ResponseEntity<?> handleException(Exception e, int status) {
        return ResponseEntity
                .status(status)
                .body(
                        HTTPErrorResponse.builder()
                                .message(e.getMessage())
                                .status(status)
                                .build()
                );
    }
}

File: backend/src/main/kotlin/com/backend/ord/exceptions/handlers/RESTExceptionHandler.kt
package com.backend.ord.exceptions.handlers

import com.backend.ord.api.responses.HTTPErrorResponse
import com.backend.ord.exceptions.REST.*
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.annotation.ControllerAdvice
import org.springframework.web.bind.annotation.ExceptionHandler


@ControllerAdvice
class RESTExceptionHandler {
    @ExceptionHandler(
        BadRequestException::class,
        UnauthorizedException::class,
        ForbiddenException::class,
        NotFoundException::class,
        PayloadTooLargeException::class,
        InternalServerError::class
    )
    fun handleException(e: Exception): ResponseEntity<HTTPErrorResponse> {
        val status = getStatusForException(e)
        val errorResponse = HTTPErrorResponse(message = e.message, status = status)

        return ResponseEntity.status(status).body(errorResponse)
    }

    // Catch-all exception handler
    @ExceptionHandler(Exception::class)
    fun handleUncaughtException(e: Exception): ResponseEntity<HTTPErrorResponse> {
        // TODO: Improve error logging
        val errorResponse = HTTPErrorResponse(
            message = """
                Unexpected error occurred
                - Exception: $e
                - Message: ${e.message}
                - Cause: ${e.cause}
                - Stack trace: ${e.stackTrace}
                - Class: ${e.javaClass}
                - Localized message: ${e.localizedMessage}
                - Suppressed: ${e.suppressed}
            """,

            status = HttpStatus.INTERNAL_SERVER_ERROR.value()
        )

        return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(errorResponse)
    }

    @ExceptionHandler(MethodArgumentNotValidException::class)
    fun handleMethodArgumentNotValidException(e: MethodArgumentNotValidException): ResponseEntity<HTTPErrorResponse> {
        val status = HttpStatus.BAD_REQUEST.value()
        val errorResponse = HTTPErrorResponse(
            message = e.bindingResult.allErrors.joinToString { it.defaultMessage ?: "" },
            status = status
        )

        return ResponseEntity.status(status).body(errorResponse)
    }

    private fun getStatusForException(e: Exception): Int = when (e) {
        is BadRequestException -> HttpStatus.BAD_REQUEST.value() // 400
        is UnauthorizedException -> HttpStatus.UNAUTHORIZED.value() // 401
        is ForbiddenException -> HttpStatus.FORBIDDEN.value() // 403
        is NotFoundException -> HttpStatus.NOT_FOUND.value() // 404
        is PayloadTooLargeException -> HttpStatus.PAYLOAD_TOO_LARGE.value() // 413
        is InternalServerError -> HttpStatus.INTERNAL_SERVER_ERROR.value() // 500
        is BadGatewayException -> HttpStatus.BAD_GATEWAY.value() // 502
        else -> HttpStatus.INTERNAL_SERVER_ERROR.value() // 500
    }
}
